File: snippets/go/metadata.go
package example

import (
	"encoding/json"
	"log"

	"github.com/breez/breez-sdk-go/breez_sdk"
)

func SetPaymentMetadata() error {
	// ANCHOR: set-payment-metadata
	err := sdk.SetPaymentMetadata("target-payment-hash", `{"myCustomValue":true}`)
	if err != nil {
		return err
	}
	// ANCHOR_END: set-payment-metadata
	return nil
}

func FilterPaymentMetadata() error {
	// ANCHOR: filter-payment-metadata
	metadataFilters := []breez_sdk.MetadataFilter{
		{JsonPath: "myCustomValue", JsonValue: "true"},
	}

	payments, err := sdk.ListPayments(breez_sdk.ListPaymentsRequest{
		MetadataFilters: &metadataFilters,
	})

	if err != nil {
		return err
	}
	// ANCHOR_END: filter-payment-metadata
	log.Printf("%#v", payments)
	return nil
}

func FilterPaymentMetadataString() {
	// ANCHOR: filter-payment-metadata-string
	metadataFilters := []breez_sdk.MetadataFilter{
		{JsonPath: "customerName", JsonValue: "\"Satoshi Nakamoto\""},
	}

	jsonValue, _ := json.Marshal("Satoshi Nakamoto")
	metadataFilters = []breez_sdk.MetadataFilter{
		{
			JsonPath:  "customerName",
			JsonValue: string(jsonValue),
		},
	}
	// ANCHOR_END: filter-payment-metadata-string
	log.Printf("%#v", metadataFilters)
}

func FilterPaymentMetadataObject() {
	// ANCHOR: filter-payment-metadata-object
	// This will *NOT* work
	metadataFilters := []breez_sdk.MetadataFilter{
		{JsonPath: "parent.nestedArray", JsonValue: "[1, 2, 3]"},
	}

	// Any of these will work
	jsonValue, _ := json.Marshal([]int{1, 2, 3})

	metadataFilters = []breez_sdk.MetadataFilter{
		{JsonPath: "parent.nestedArray", JsonValue: "[1,2,3]"},
		{JsonPath: "parent.nestedArray", JsonValue: string(jsonValue)},
	}
	// ANCHOR_END: filter-payment-metadata-object
	log.Printf("%#v", metadataFilters)
}

File: snippets/rust/src/metadata.rs
use std::sync::Arc;
use anyhow::Result;
use breez_sdk_core::*;

async fn set_payment_metadata(sdk: Arc<BreezServices>) -> Result<()> {
  // ANCHOR: set-payment-metadata
    sdk.set_payment_metadata("target-payment-hash", r#"{"myCustomValue":true}"#).await?;
  // ANCHOR_END: set-payment-metadata
  
    Ok(())
}

async fn filter_payment_metadata(sdk: Arc<BreezServices>) -> Result<()> {
  // ANCHOR: filter-payment-metadata
    let metadata_filters = vec![
        MetadataFilter {
            json_path: "myCustomValue".to_string(),
            json_value: "true".to_string(),
        },
    ];

    sdk.list_payments(ListPaymentsRequest {
        metadata_filters,
        ..Default::default(),
    }).await?;
  // ANCHOR_END: filter-payment-metadata
}

async fn filter_payment_metadata_string(sdk: Arc<BreezServices>) -> Result<()> {
  // ANCHOR: filter-payment-metadata-string
    // Note: The following are equivalent
    let metadata_filters = vec![
        MetadataFilter {
            json_path: "customerName".to_string(),
            json_value: r#""Satoshi Nakamoto""#.to_string(),
        },
        MetadataFilter {
            json_path: "customerName".to_string(),
            json_value: serde_json::json!("Satoshi Nakamoto").to_string(),
        },
    ];
  // ANCHOR_END: filter-payment-metadata-string
}

async fn filter_payment_metadata_object(sdk: Arc<BreezServices>) -> Result<()> {
  // ANCHOR: filter-payment-metadata-object
    // This will *NOT* work
    let metadata_filters = vec![
        MetadataFilter {
            json_path: "parent.nestedArray".to_string(),
            json_value: r#"[1, 2, 3]"#.to_string(),
        },
    ];

    // This will work
    let metadata_filters = vec![
        MetadataFilter {
            json_path: "parent.nestedArray".to_string(),
            json_value: r#"[1,2,3]"#.to_string(),
        },
    ];
  // ANCHOR_END: filter-payment-metadata-object
  
  // ANCHOR: filter-payment-metadata-object-serde
    let metadata_filters = vec![
        MetadataFilter {
            json_path: "parent.nestedArray".to_string(),
            json_value: serde_json::json!(&[1, 2, 3]).to_string(),
        },
    ];
  // ANCHOR_END: filter-payment-metadata-object-serde
    Ok(())
}
